modin/experimental/batch/pipeline.py

```diff
@@ -0,0 +1,386 @@
+    this query, and combines them into 1 partition.
+
+    Notes
+    -----
+    `func` must be a function that is applied along an axis of the dataframe.
+
+    Use `pandas` for any module level functions inside `func` since it operates directly on
+    partitions.
+    """
+
+    def __init__(
+        self,
+        func: Callable,
+        is_output: bool = False,
+        repartition_after: bool = False,
+        fan_out: bool = False,
+        pass_partition_id: bool = False,
+        reduce_fn: Optional[Callable] = None,
+    ):
+        self.func = func
+        self.is_output = is_output
+        self.fan_out = fan_out
+        self.repartition_after = repartition_after
+        self.pass_partition_id = pass_partition_id
+        # List of sub-queries to feed into this query, if this query is an output node.
+        self.operators = None
+        self.reduce_fn = reduce_fn
+
+
+class PandasQueryPipeline(object):
+    """
+    Internal representation of a query pipeline.
+
+    This object keeps track of the functions that compose to form a query pipeline.
+
+    Parameters
+    ----------
+    df : modin.pandas.Dataframe
+        The dataframe to perform this pipeline on.
+    num_partitions : int, optional
+        The number of partitions to maintain for the batched dataframe.
+        If not specified, the value is assumed equal to ``NPartitions.get()``.
+
+    Notes
+    -----
+    Only row-parallel pipelines are supported. All queries will be applied along the row axis.
+    """
+
+    def __init__(self, df, num_partitions: int = -1):
+        if Engine.get() != "Ray" or (
+            not isinstance(df._query_compiler._modin_frame, PandasOnRayDataframe)
+        ):  # pragma: no cover
+            ErrorMessage.not_implemented(
+                "Batch Pipeline API is only implemented for Ray Engine."
+            )
+        ErrorMessage.single_warning(
+            "The Batch Pipeline API is an experimental feature and still under development in Modin."
+        )
+        self.df = df
+        self.num_partitions = (
+            num_partitions if num_partitions > 0 else NPartitions.get()
+        )
+        self.outputs = []  # List of output queries.
+        self.nodes_list = []  # List of all queries.
+        self.node_to_id = (
+            None  # Optional mapping of node to output ID (if specified in `add_query`).
+        )
+
+    def update_df(self, df):
+        """
+        Update the dataframe to perform this pipeline on.
+
+        Parameters
+        ----------
+        df : modin.pandas.DataFrame
+            The new dataframe to perform this pipeline on.
+        """
+        if Engine.get() != "Ray" or (
+            not isinstance(df._query_compiler._modin_frame, PandasOnRayDataframe)
+        ):  # pragma: no cover
+            ErrorMessage.not_implemented(
+                "Batch Pipeline API is only implemented for Ray Engine."
+            )
```

Comment:
Why do we need these warnings outside of the` __init__`?

Reply:
This is in `update_df` - we want to make sure that we do the same checks when we update the dataframe that this pipeline will run on.